src/goto-programs/remove_function_pointers.cpp

```diff
@@ -140,6 +162,317 @@ symbolt &remove_function_pointerst::new_tmp_symbol()
+          }
+          else if(try_get_call_from_symbol(op, out_functions))
+          {
+          }
+          else
+          {
+            // return false?
+            // in this case there is an element
+          }
+        }
+        return out_functions.size() > 0;
+      }
+    }
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_function_pointerst::try_evaluate_index_value
+
+  Inputs:
+   index_expr - The expression of the index of the index expression (e.g.
+                index_exprt::index())
+
+ Outputs: Returns true if was able to find a constant value for the index
+          expression. If true, then out_array_index will be the index within
+          the array that the function pointer is pointing to.
+
+ Purpose: Given an index into an array, resolve, if possible, the index
+          that is being accessed. This deals with symbols and typecasts to
+          constant values.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_evaluate_index_value(
+  const exprt &index_value_expr, mp_integer &out_array_index)
+{
+  if(index_value_expr.id()==ID_constant)
+  {
+    constant_exprt constant_expr=to_constant_expr(index_value_expr);
+    mp_integer array_index;
+    bool errors=to_integer(constant_expr, array_index);
+    if(!errors)
+    {
+      out_array_index=array_index;
+    }
+    return !errors;
+  }
+  else if(index_value_expr.id()==ID_symbol)
+  {
+    // Resolve symbol and try again
+    const symbolt &index_symbol=
+      symbol_table.lookup(index_value_expr.get(ID_identifier));
+
+    return try_evaluate_index_value(index_symbol.value, out_array_index);
+  }
+  else if(index_value_expr.id()==ID_typecast)
+  {
+    // Follow the typecast
+
+    return try_evaluate_index_value(index_value_expr.op0(), out_array_index);
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_function_pointerst::try_get_array_from_index
+
+  Inputs:
+   index_expr - The expression of the index in the array
+
+ Outputs: Returns true if was able to find the array associated
+          with the index. If true, then out_array_expr will contain
+          the expression representing the array.
+
+ Purpose: Given an index into an array, resolve the array expression,
+          following symbol expressions through to the actual array
+          expression.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_get_array_from_index(
+  const index_exprt &index_expr, array_exprt &out_array_expr)
+{
+  exprt array_expr=index_expr.array();
+  if(array_expr.id()==ID_array)
+  {
+    out_array_expr=to_array_expr(array_expr);
+    return true;
+  }
+  else if(array_expr.id()==ID_symbol)
+  {
+    // Is there some existing code to follow symbols to the real value?
```

Comment:
Sorry I'm not sure I fully understand - am I not already following the symbols to their values? I'm currently (recursively) when I find a `ID_symbol`, look at its `ID_identifier`, look that up in the `symbol_table` and look at the symbols `value`. 

I was hoping this logic for recursively resolving the symbols (and ideally dealing with other things like casting) would exist in the `simplify_expr` but it doesn't seem to be working as I had hoped.